server/src/main/java/org/elasticsearch/index/search/MultiMatchQueryParser.java

```diff
@@ -106,15 +99,216 @@ private Query combineGrouped(List<Query> groupQuery, float tieBreaker) {
                 queries.add(query);
             }
         }
-        return queries;
+        return combineWithDisMax(queries, tieBreaker);
     }
 
-    private List<Query> buildCrossFieldQuery(Map<String, Float> fieldNames,
-                                             Object value, String minimumShouldMatch, float tieBreaker) {
+    private Query combineWithDisMax(List<Query> groupQuery, float tieBreaker) {
+        if (groupQuery.isEmpty()) {
+            return zeroTermsQuery();
+        }
+        if (groupQuery.size() == 1) {
+            return groupQuery.get(0);
+        }
+        return new DisjunctionMaxQuery(groupQuery, tieBreaker);
+    }
+
+    private Query buildCombinedFieldQuery(Map<String, Float> fields, Object value, String minimumShouldMatch) {
+        validateSimilarity(fields);
 
+        List<Query> queries = new ArrayList<>();
+        Map<Analyzer, List<FieldAndBoost>> groups = new HashMap<>();
```

Comment:
When writing tests, I noticed that this analyzer grouping is slightly buggy -- when using `normalizer` plus `split_queries_on_whitespace`, we group keyword fields together with different analyzer definitions. This bug seems to already exist for `cross_fields`. If we stick with analyzer grouping we'll need to address this bug.